### PR TITLE
SDI-360 Update Nomis from incentive creation domain event

### PR DESCRIPTION
**Pull request by steverendell**
Adding tests

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesMappingService.kt [hunk 1]
```diff
@@ -10,45 +10,15 @@ import reactor.core.publisher.Mono
 import java.time.LocalDateTime
 
 @Service
-class MappingService(
+class IncentivesMappingService(
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesMappingService.kt:13]: split to separate classes

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/listeners/PrisonerDomainEventsListener.kt [hunk 2]
```diff
@@ -34,67 +35,43 @@ class PrisonerDomainEventsListener(
...
-        telemetryClient.trackEvent(
-          "incentive-retry-received",
-          mapOf(
-            "id" to context.incentiveId.toString(),
-            "nomisBookingId" to context.nomisBookingId.toString(),
-            "nomisIncentiveSequence" to context.nomisIncentiveSequence.toString()
-          ),
-        )
-        incentivesService.createIncentiveRetry(context)
-      }
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/listeners/PrisonerDomainEventsListener.kt:85]: rearranged these when I realised both visit nd incentive messages will have the same 'type' but different 'eventType'

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/services/UpdateQueueService.kt [hunk 3]
```diff
@@ -19,42 +21,48 @@ class UpdateQueueService(
...
+  property = "type",
+  defaultImpl = VisitContext::class
+)
+@JsonSubTypes(
+  JsonSubTypes.Type(value = VisitContext::class, name = "VISIT"),
+  JsonSubTypes.Type(value = IncentiveContext::class, name = "INCENTIVE")
+)
+abstract class Context(val type: String) {
+  abstract fun getId(): String
+}
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/services/UpdateQueueService.kt:53]: This allows jackson to automatically deserialise the retry message to the right context subclass

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/helpers/Messages.kt [hunk 4]
```diff
@@ -31,7 +31,29 @@ fun prisonVisitCancelledMessage(visitId: String = "12", prisonerId: String = "AB
 fun retryMessage() = """
       {
         "Type":"RETRY",
-        "Message":"{\"nomisId\":\"12345\",\"vsipId\":\"12\"}",
+        "Message":"{\"type\":\"VISIT\",\"nomisId\":\"12345\",\"vsipId\":\"12\"}",
```
Comment (steverendell) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/helpers/Messages.kt:34]: 'type' here tells jackson which class to deserialise to

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/wiremock/MappingMockServer.kt [hunk 5]
```diff
@@ -110,6 +110,71 @@ class MappingMockServer : WireMockServer(WIREMOCK_PORT) {
     )
   }
```
Comment (steverendell) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/wiremock/MappingMockServer.kt:111]: unfortunately i couldnt split this into 2 classes as you cant have 2 Mockservers on the same port

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesService.kt [hunk 6]
```diff
@@ -49,7 +47,7 @@ class IncentivesService(
         "iepTime" to iepTime.format(DateTimeFormatter.ISO_TIME),
       )
 
-      if (mappingService.getIncentiveMappingGivenIncentiveId(event.incentiveId) != null) {
+      if (mappingService.getMappingGivenIncentiveId(event.incentiveId) != null) {
         telemetryClient.trackEvent("incentive-get-map-failed", telemetryMap)
         log.warn("Mapping already exists for incentive id $event.id")
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/incentives/IncentivesService.kt:52]: is the event more of a "incentive-create-map-failed"?

**Review by lbennett-moj — APPROVED** [review]

**Pull request merged**